src/coreclr/tools/Common/Compiler/TypeWithRepeatedFieldsFieldLayoutAlgorithm.cs

```diff
@@ -0,0 +1,78 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Internal.TypeSystem;
+
+namespace ILCompiler
+{
+    /// <summary>
+    /// Represents an algorithm that computes field layout for intrinsic vector types (Vector64/Vector128/Vector256).
+    /// </summary>
+    public class TypeWithRepeatedFieldsFieldLayoutAlgorithm : FieldLayoutAlgorithm
+    {
+        private readonly FieldLayoutAlgorithm _fallbackAlgorithm;
+
+        public TypeWithRepeatedFieldsFieldLayoutAlgorithm(FieldLayoutAlgorithm fallbackAlgorithm)
+        {
+            _fallbackAlgorithm = fallbackAlgorithm;
+        }
+
+        public override ComputedInstanceFieldLayout ComputeInstanceLayout(DefType defType, InstanceLayoutKind layoutKind)
+        {
+            var type = (TypeWithRepeatedFields)defType;
+
+            ComputedInstanceFieldLayout layoutFromMetadata = _fallbackAlgorithm.ComputeInstanceLayout(type.MetadataType, layoutKind);
+
+            FieldAndOffset[] offsets = layoutFromMetadata.Offsets;
+
+            if (offsets is not null)
+            {
+                var fieldOffsets = new FieldAndOffset[type.NumFields];
+
+                LayoutInt cumulativeOffset = new LayoutInt(0);
+                int fieldIndex = 0;
+                foreach (FieldDesc field in type.GetFields())
+                {
+                    if (field.IsStatic)
+                    {
+                        continue;
+                    }
+                    fieldOffsets[fieldIndex++] = new FieldAndOffset(field, cumulativeOffset);
+                    cumulativeOffset += field.FieldType.GetElementSize();
+                }
+
+                offsets = fieldOffsets;
+            }
+
+            return new ComputedInstanceFieldLayout
+            {
+                ByteCountUnaligned = layoutFromMetadata.ByteCountUnaligned,
+                ByteCountAlignment = layoutFromMetadata.ByteCountAlignment,
+                FieldAlignment = layoutFromMetadata.FieldAlignment,
+                FieldSize = layoutFromMetadata.FieldSize,
+                Offsets = offsets,
+                LayoutAbiStable = layoutFromMetadata.LayoutAbiStable
+            };
+        }
+
+        public override ComputedStaticFieldLayout ComputeStaticFieldLayout(DefType defType, StaticLayoutKind layoutKind)
+        {
+            return _fallbackAlgorithm.ComputeStaticFieldLayout(((TypeWithRepeatedFields)defType).MetadataType, layoutKind);
+        }
+
+        public override bool ComputeContainsGCPointers(DefType type)
+        {
+            return _fallbackAlgorithm.ComputeContainsGCPointers(((TypeWithRepeatedFields)type).MetadataType);
```

Comment:
A couple of these give me a pause - would we expect to ever be here if `ComputeContainsGCPointers` returns true? Similarly if `ComputeIsUnsafeValueType` returns false.

Right now we probably would. The question is if it's expected/makes sense?

Reply:
Yes, we could have `ComputeContainsGCPointers` return true if the struct is (for example) an `InlineArray` struct of strings.

I believe we could have `ComputeIsUnsafeValueType` return true if you do an `InlineArray` of a traditional fixed buffer, like so: https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEAMACdBGAOgCUBXGAFyjAFMCBhAezAAcoAbagJwGUuA3KAGNqAZwDcAWABQ0gNoBJGG1jUAgp04QAngAoAzAEoAutPR7sZERABm1bCPKcSg8tm7SA3tOw/c561AAHtQI2LCugbIALEaSUgC+QA